Docs: Add warning against installing Ray in base environment (#3267)
skypilot-org/skypilot: docs/source/running-jobs/distributed-jobs.rst

  # Worker nodes.
  $ ssh mycluster-worker1
  $ ssh mycluster-worker2

> Executing a Distributed Ray Program

Reviewer: For discussion: We may want to add a dedicated guide on how to use Ray as part of the user job. That page can use a tip like this. A full YAML example https://github.com/skypilot-org/skypilot/issues/3195 may help promote this recommendation more directly.

Reviewer 2: Yes, let’s add a yaml example of starting a distributed job with ray as mentioned above. We can try to move it to a separate page later after this pr is merged.

Author: So for confirmation: I should add the command to run the example in distributed_ray_train, the one that uses FashionMNIST and add the command for tha yaml file and its output as example in the documentation ?

Reviewer 2: Yes, we can add the yaml file inline with the command to run that [yaml](https://github.com/skypilot-org/skypilot/blob/master/examples/distributed_ray_train/ray_train.yaml) : )

  # Worker nodes.
  $ ssh mycluster-worker1
  $ ssh mycluster-worker2

Executing a Distributed Ray Program
------------------------------------

> To execute a distributed Ray program on many VMs, you can use the following example:

Reviewer 2: ```suggestion
To execute a distributed Ray program on many VMs, you can download the `training script <https://github.com/skypilot-org/skypilot/blob/master/examples/distributed_ray_train/train.py>`_ and launch the `task yaml <https://github.com/skypilot-org/skypilot/blob/master/examples/distributed_ray_train/ray_train.yaml>`_:
```

.. code-block:: console

> 
>   $ sky launch ray_train.yaml

Reviewer 2: ```suggestion

  $ wget https://raw.githubusercontent.com/skypilot-org/skypilot/master/examples/distributed_ray_train/train.py
  $ sky launch ray_train.yaml
```

.. code-block:: yaml

>     :emphasize-lines: 6-6,21-22,24-25

Reviewer 2: It seems the emphasize lines seems unrelated here. Should we just remove it?
```suggestion
```

  
    resources:
      accelerators: L4:2
      memory: 64+
  
    num_nodes: 2

>   
>     setup: |

Reviewer 2: ```suggestion

    workdir: .

    setup: |
```

>       pip install "ray[train]"

Reviewer 2: Let's keep the original conda commands to follow the warning below.
```suggestion
      conda activate ray
      if [ $? -ne 0 ]; then
        conda create -n ray python=3.10 -y
        conda activate ray
      fi
      
      pip install "ray[train]"
```

      pip install tqdm
      pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
  
    run: |
      sudo chmod 777 -R /var/tmp
      head_ip=`echo "$SKYPILOT_NODE_IPS" | head -n1`
      num_nodes=`echo "$SKYPILOT_NODE_IPS" | wc -l`
      ray start --head  --disable-usage-stats --port 6379
      if [ "$SKYPILOT_NODE_RANK" == "0" ]; then
        ps aux | grep ray | grep 6379 &> /dev/null || ray start --head  --disable-usage-stats --port 6379
        sleep 5
        python train.py --num-workers $num_nodes
      else
        sleep 5
        ps aux | grep ray | grep 6379 &> /dev/null || ray start --address $head_ip:6379 --disable-usage-stats
      fi

.. warning:: 

>   **Avoid Installing Ray in Base Environment**
> 
>   Before proceeding with the execution of a distributed Ray program, it is crucial to ensure that Ray is **not** installed in the base environment. Installing a different version of Ray in the base environment can lead to compatibility issues, conflicts, and unintended consequences.
> 
>   To maintain a clean and stable environment for your distributed Ray program, it is highly recommended to **create a dedicated virtual environment** for Ray and its dependencies. This helps isolate the Ray installation and prevents interference with other packages in your base environment.

Reviewer 2: ```suggestion
  **Avoid Installing Ray in Base Environment**: Before proceeding with the execution of a distributed Ray program, it is crucial to ensure that Ray is **not** installed in the *base* environment. Installing a different version of Ray in the base environment can lead to abnormal cluster status.

  It is highly recommended to **create a dedicated virtual environment** (as above) for Ray and its dependencies, and avoid calling `ray stop` as that will also cause issue with the cluster.
```

  # Worker nodes.
  $ ssh mycluster-worker1
  $ ssh mycluster-worker2

Executing a Distributed Ray Program
------------------------------------
To execute a distributed Ray program on many VMs, you can download the `training script <https://github.com/skypilot-org/skypilot/blob/master/examples/distributed_ray_train/train.py>`_ and launch the `task yaml <https://github.com/skypilot-org/skypilot/blob/master/examples/distributed_ray_train/ray_train.yaml>`_:

.. code-block:: console

  $ wget https://raw.githubusercontent.com/skypilot-org/skypilot/master/examples/distributed_ray_train/train.py
  $ sky launch ray_train.yaml

.. code-block:: yaml
  
    resources:
      accelerators: L4:2
      memory: 64+
  
    num_nodes: 2

    workdir: .

    setup: |

>       conda activate ray
>       if [ $? -ne 0 ]; then
>         conda create -n ray python=3.10 -y
>         conda activate ray
>       fi
>       

Reviewer 2: ```suggestion
```

Duplicated?

To execute a distributed Ray program on many VMs, you can download the `training script <https://github.com/skypilot-org/skypilot/blob/master/examples/distributed_ray_train/train.py>`_ and launch the `task yaml <https://github.com/skypilot-org/skypilot/blob/master/examples/distributed_ray_train/ray_train.yaml>`_:

.. code-block:: console

  $ wget https://raw.githubusercontent.com/skypilot-org/skypilot/master/examples/distributed_ray_train/train.py
  $ sky launch ray_train.yaml

.. code-block:: yaml
  
    resources:
      accelerators: L4:2
      memory: 64+
  
    num_nodes: 2

    workdir: .

    setup: |
      conda activate ray
      if [ $? -ne 0 ]; then
        conda create -n ray python=3.10 -y
        conda activate ray
      fi
      
      conda activate ray
      if [ $? -ne 0 ]; then
        conda create -n ray python=3.10 -y
        conda activate ray
      fi
      
      pip install "ray[train]"
      pip install tqdm
      pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
  
    run: |
      sudo chmod 777 -R /var/tmp
      head_ip=`echo "$SKYPILOT_NODE_IPS" | head -n1`
      num_nodes=`echo "$SKYPILOT_NODE_IPS" | wc -l`

>       ray start --head  --disable-usage-stats --port 6379

Reviewer 2: This line shouldbe removed : )
```suggestion
```
